### PR TITLE
[CPU] Fix issues when running llama3.2-11B vision model with image tasks

### PR DESCRIPTION
**Pull request by blzheng**
## Motivation

This PR fixes several issues encountered when running the LLaMA 3.2 11B vision model with image tasks on CPU.

## Modifications

- Fixed kv cache access logic for cross attention
  - Affected both backends: 
      - python/sglang/srt/layers/attention/torch_native_backend.py 
      - python/sglang/srt/layers/attention/intel_amx_backend.py
  - Updated the CPU kernel to support such case:
      -  sgl-kernel/csrc/cpu/extend.cpp
      -  sgl-kernel/csrc/cpu/decode.cpp
      -  sgl-kernel/csrc/cpu/torch_extension_cpu.cpp
- Removed hard-coded CUDA device in python/sglang/srt/models/mllama.py
- Fixed incorrect attribute access of mm_inputs in python/sglang/srt/models/mllama.py
- Fixed unrecognized chat template parsing in python/sglang/bench_serving.py
- Supported >2D inputs for weight_packed_linear in python/sglang/srt/layers/quantization/unquant.py

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

---
### CI States

Latest PR Test (Base): :x: [Run #26202393853](https://github.com/sgl-project/sglang/actions/runs/26202393853)
Latest PR Test (Extra): :x: [Run #26202393811](https://github.com/sgl-project/sglang/actions/runs/26202393811)

### CODE REVIEW

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 1]
```diff
@@ -49,6 +50,183 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+                    enable_gqa=enable_gqa,
+                    scale=scaling,
+                    is_causal=causal,
+                )
+                .squeeze(0)
+                .movedim(query.dim() - 2, 0)
+            )
+            output[start_q:end_q, :, :] = per_req_out_redudant[prefill_seq_len_q:, :, :]
+            start_q, start_kv = end_q, end_kv
+        return output
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py]: My understanding is that this PR implements torch implementation for both `extend` and  `decode` under the backend of `intel_amx_backend`?

This is not very good practice. I recommend:
* upgrade intel_amx_backend to support vl models natively
* upgrade torch native backend to support vl models as well (non-amx path)

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
@blzheng you can summarize the topic a little bit and we can discuss internally, what we need to know:

* difference than existing model, e.g. K is none or whatever.
* input shapes and dtypes for each tensor in the targeted model or model(s).

File: python/sglang/srt/layers/attention/intel_amx_backend.py (hunk 1 above)
Comment (blzheng) [reply python/sglang/srt/layers/attention/intel_amx_backend.py]: I plan to first use SDPA (via PyTorch's Flash Attention) for the cross-attention part in `intel_amx_backend`, while keeping the rest on the existing `extend` and `decode` kernels. A follow-up PR will add native cross-attention support to those kernels. Does this approach work for you?

Comment (blzheng) [reply python/sglang/srt/layers/attention/intel_amx_backend.py]: I have updated both extend and decode to support cross attention in this model. Please help review again. Thanks~

File: python/sglang/bench_serving.py [hunk 2]
```diff
@@ -879,22 +880,37 @@ def sample_mmmu_requests(
...
+                                                "type": "image_url",
+                                                "image_url": {"url": image_data},
+                                            },
+                                            {"type": "text", "text": prompt},
+                                        ],
+                                    }
+                                ],
+                                add_generation_prompt=True,
+                                tokenize=False,
+                            )
```
Comment (mingfeima) [inline python/sglang/bench_serving.py]:
```suggestion
                        if "llama" in tokenizer.name_or_path.lower():
                            image_field = {"type": "image", "image": image_data}
                        else:
                            image_field = {"type": "image_url", "image_url": {"url": image_data}}

                       prompt = tokenizer.apply_chat_template(
                            [
                                 {
                                     "role": "user",
                                      "content": [
                                          image_field,
                                          {"type": "text", "text": prompt},
                                       ],
                                   }
                               ],
                              add_generation_prompt=True,
                              tokenize=False,
                          )
```

try remove the duplicated code.

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 3]
```diff
@@ -62,14 +62,17 @@ def forward_extend(
...
+            if not layer.is_cross_attention
+            else forward_batch.encoder_out_cache_loc
+        )
         if save_kv_cache:
-            forward_batch.token_to_kv_pool.set_kv_buffer(
-                layer, forward_batch.out_cache_loc, k, v
-            )
+            if k is not None:
+                assert v is not None
+                forward_batch.token_to_kv_pool.set_kv_buffer(layer, cache_loc, k, v)
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py]:
```
if save_kv_cache and k is not None and v is not None:
    forward_batch.token_to_kv_pool.set_kv_buffer(layer, cache_loc, k, v)
```

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 4]
```diff
@@ -62,14 +62,17 @@ def forward_extend(
             o = q.new_empty((q.shape[0], layer.tp_q_head_num * layer.v_head_dim))
         else:
             o = torch.empty_like(q)
-
+        cache_loc = (
+            forward_batch.out_cache_loc
+            if not layer.is_cross_attention
+            else forward_batch.encoder_out_cache_loc
+        )
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py:106]: is this a runtime constant, if yet we can move `cache_loc` to `self.forward_metadata`, which means it will be only executed once, not for every iteration.

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 5]
```diff
@@ -105,23 +110,28 @@ def forward_decode(
...
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py:152]: same as above.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 6]
```diff
@@ -35,9 +35,11 @@ def _run_sdpa_forward_extend(
         seq_lens: torch.Tensor,
         extend_prefix_lens: torch.Tensor,
         extend_seq_lens: torch.Tensor,
+        encoder_lens=None,
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py]:
```suggestion
        encoder_lens: Optional[torch.Tensor] = None,
```

Use Optional.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 7]
```diff
@@ -75,8 +79,14 @@ def _run_sdpa_forward_extend(
...
-            end_kv = start_kv + seq_len_kv
-
+            if encoder_lens is not None:
+                start_kv = 0 if is_cross_attn else encoder_lens[seq_idx]
+                end_kv = (
+                    encoder_lens[seq_idx] if is_cross_attn else start_kv + seq_len_kv
+                )
+            else:
+                start_kv = 0
+                end_kv = start_kv + seq_len_kv
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py:91]:
```suggestion
if encoder_lens is not None:
    if is_cross_attn:
        start_kv = 0
        end_kv = encoder_lens[seq_idx]
    else:
        start_kv = encoder_lens[seq_idx]
        end_kv = start_kv + seq_len_kv
else:
    start_kv = 0
    end_kv = seq_len_kv
```

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 8]
```diff
@@ -151,14 +165,22 @@ def _run_sdpa_forward_decode(
...
             end_q = start_q + seq_len_q
-            end_kv = start_kv + seq_len_kv
+            if encoder_lens is not None:
+                start_kv = 0 if is_cross_attn else encoder_lens[seq_idx]
+                end_kv = (
+                    encoder_lens[seq_idx] if is_cross_attn else start_kv + seq_len_kv
+                )
+            else:
+                start_kv = 0
+                end_kv = start_kv + seq_len_kv
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py:184]: same as above.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 9]
```diff
@@ -192,11 +214,15 @@ def forward_extend(
             o = q.new_empty((q.shape[0], layer.tp_q_head_num * layer.v_head_dim))
         else:
             o = torch.empty_like(q)
-
+        cache_loc = (
+            forward_batch.out_cache_loc
+            if not layer.is_cross_attention
+            else forward_batch.encoder_out_cache_loc
+        )
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py]: same as above

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 10]
```diff
@@ -192,11 +214,15 @@ def forward_extend(
...
         if save_kv_cache:
-            forward_batch.token_to_kv_pool.set_kv_buffer(
-                layer, forward_batch.out_cache_loc, k, v
-            )
+            if k is not None:
+                assert v is not None
+                forward_batch.token_to_kv_pool.set_kv_buffer(layer, cache_loc, k, v)
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py]: same as above.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 11]
```diff
@@ -217,9 +243,11 @@ def forward_extend(
             forward_batch.seq_lens,
             forward_batch.extend_prefix_lens,
             forward_batch.extend_seq_lens,
+            encoder_lens=forward_batch.encoder_lens,
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py]:
```suggestion
            forward_batch.encoder_lens,
```

I remember intel python team suggests not use '=' if possible, it is slower a little bit.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 12]
```diff
@@ -240,11 +268,16 @@ def forward_decode(
             o = q.new_empty((q.shape[0], layer.tp_q_head_num * layer.v_head_dim))
         else:
             o = torch.empty_like(q)
+        cache_loc = (
+            forward_batch.out_cache_loc
+            if not layer.is_cross_attention
+            else forward_batch.encoder_out_cache_loc
+        )
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py:288]: same as above.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 13]
```diff
@@ -259,9 +292,11 @@ def forward_decode(
             forward_batch.req_to_token_pool.req_to_token,
             forward_batch.req_pool_indices,
             forward_batch.seq_lens,
+            encoder_lens=forward_batch.encoder_lens,
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/torch_native_backend.py]: same as above

File: python/sglang/srt/layers/linear.py [hunk 14]
```diff
@@ -196,7 +196,7 @@ def __init__(
 
         if bias:
             self.bias = Parameter(
-                torch.empty(self.output_size, dtype=self.params_dtype)
+                torch.zeros(self.output_size, dtype=self.params_dtype)
```
Comment (mingfeima) [inline python/sglang/srt/layers/linear.py]: why we need this?

`empty` is lazy initialization, it is much faster.

File: python/sglang/srt/layers/quantization/unquant.py [hunk 15]
```diff
@@ -116,9 +116,18 @@ def apply(
...
+                x = x.view(-1, layer.weight.shape[-1])
+            output = torch.ops.sgl_kernel.weight_packed_linear(
                 x, layer.weight, bias, True  # is_vnni
             )
+            if len(x_shapes) >= 3:
+                if x_shapes[-1] == layer.weight.shape[-1]:
+                    output = output.view(*x_shapes[:-1], -1)
+                else:
+                    output = output.view(*x_shapes[:-2], layer.weight.shape[0])
+            return output
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/unquant.py:156]: fix this is python is going to create a lot of python/torch overhead.

i prefer fix it in C++:
* make input: [*, K]
* make output: [*, K]

where * can be 1d to nd.

just do something in C++:

```
const int64_t ndim = input.ndimensions();
const auto input_sizes = input.sizes();
int64_t K = input_sizes[ndim - 1];
int64_t M = input.numel() / K;
int64_t N = weight.size(0);

// get output sizes
input_sizes[ndim - 1] = N;


```

**Review by mingfeima — CHANGES_REQUESTED** [review]
generally OK. just some minor changes needed.

File: python/sglang/srt/layers/attention/intel_amx_backend.py (hunk 4 above)
Comment (blzheng) [reply python/sglang/srt/layers/attention/intel_amx_backend.py:106]: It is not a runtime constant because it depends on information from the `layer`, which is provided at runtime.

File: python/sglang/srt/layers/linear.py (hunk 14 above)
Comment (blzheng) [reply python/sglang/srt/layers/linear.py]: Some biases are missing in the model weights. They should default to 0, but when initialized with `empty`, the bias may contain unexpected values, leading to accuracy issue.

File: sgl-kernel/csrc/cpu/extend.cpp [hunk 16]
```diff
@@ -433,119 +437,122 @@ void extend_attention_kernel_impl(
...
-
-        // apply causal mask
-        if (num_keys - n <= BLOCK_N) {
-          for (int row = 0; row < m_size; ++row) {
-            int last_col = m + row - n;
-            // fill [last_col + 1, n_size) to -inf
-            float* row_ptr = s_i + row * BLOCK_N;
-            fill_stub(row_ptr + last_col + 1, -std::numeric_limits<float>::infinity(), n_size - last_col - 1);
+      }
+      if (!is_cross_attn) {
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/extend.cpp:180]: so for `is_cross_attn`, we don't have stage2?

**Review by mingfeima — APPROVED** [review]

File: sgl-kernel/csrc/cpu/extend.cpp (hunk 16 above)
Comment (blzheng) [reply sgl-kernel/csrc/cpu/extend.cpp:180]: Yes

Comment (mingfeima) [reply sgl-kernel/csrc/cpu/extend.cpp:180]: shall we do 
```
if (is_cross_attn) {
  return;
}
```

save the indenting diff +- ...

Comment (blzheng) [reply sgl-kernel/csrc/cpu/extend.cpp:180]: We probably cannot return here, because the following post-processing is necessary.
https://github.com/sgl-project/sglang/blob/d08663eec1786424d33f2a7055bd03f51370e867/sgl-kernel/csrc/cpu/extend.cpp#L549-L553

**Comment by zhyncs** [comment]
@blzheng please fix the lint issue, thanks

**Comment by blzheng** [comment]
> @blzheng please fix the lint issue, thanks

Done.

File: python/sglang/srt/layers/linear.py
Comment (zhyncs) [inline python/sglang/srt/layers/linear.py:1]: why do we change to zeros here

Comment (blzheng) [reply python/sglang/srt/layers/linear.py:1]: Some biases are missing in the model weights. They should default to 0, but when initialized with empty, the bias may contain unexpected values, leading to accuracy issue.

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 17]
```diff
@@ -129,12 +143,14 @@ def _run_sdpa_forward_decode(
             output: [num_tokens, num_heads, head_size]
             k_cache: [max_total_num_tokens, num_heads, head_size]
             v_cache: [max_total_num_tokens, num_heads, head_size]
-            req_to_token: [max_num_reqs, max_context_len]
-            req_pool_indices: [num_seqs]
+            req_to_token: [max_num_reqs, max_context_len],
+            req_pool_indices: [num_seqs],
```
Comment (Alcanderian) [inline python/sglang/srt/layers/attention/torch_native_backend.py]: typo?

File: python/sglang/srt/layers/linear.py
Comment (Alcanderian) [reply python/sglang/srt/layers/linear.py:1]: > Some biases are missing in the model weights. They should default to 0, but when initialized with empty, the bias may contain unexpected values, leading to accuracy issue.

Can we tell which model?

Comment (blzheng) [reply python/sglang/srt/layers/linear.py:1]: https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct

File: python/sglang/srt/layers/attention/torch_native_backend.py (hunk 17 above)
Comment (blzheng) [reply python/sglang/srt/layers/attention/torch_native_backend.py]: Thanks. Fixed.

File: python/sglang/srt/layers/linear.py
Comment (blzheng) [reply python/sglang/srt/layers/linear.py:1]: I changed the bias initialization to use torch.zeros only on CPU. Does that make sense to you?

File: test/srt/cpu/test_decode.py [hunk 18]
```diff
@@ -18,9 +18,11 @@ def _run_sdpa_forward_decode(
         req_to_token: torch.Tensor,
         req_pool_indices: torch.Tensor,
         seq_lens: torch.Tensor,
+        encoder_lens=None,
```
Comment (mingfeima) [inline test/srt/cpu/test_decode.py:21]: No need to change `srt/cpu`, this folder will be removed later.

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Review by mingfeima — APPROVED** [review]

File: python/sglang/srt/layers/linear.py
Comment (mingfeima) [reply python/sglang/srt/layers/linear.py:1]: i am still confused why cpu need this. No worries let me check the code again.
